Missing Exception Call Stack for ExceptionInInitializerError

**Context**
 - Plugin: Infinitest for Eclipse
 - Plugin Version: 5.4.2
 - IDE: Eclipse 2024-12 (4.34.0)
 - JMockit: 1.54.0

**Bug Description**
I just installed Infinitest in Eclipse and ran into at least 3 bugs.  The first bug is being addressed in https://github.com/infinitest/infinitest/issues/430.  Once that was overcome, I found another bug.

The second bug is that Infinitest only reports `ExceptionInInitializerError` for a class.  It does not show the exception call stack and hence makes it impossible to root cause.  Here is the exception call stack.  This makes it very clear the user needs to add `-javaagent` to Infinitest's configuration file.

```
Exception (8 frames):
java.lang.ExceptionInInitializerError
	at com.mycompany.MyClassTest.<clinit>(MyClassTest.java:69)
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:375)
	at org.infinitest.testrunner.DefaultRunner.runTest(DefaultRunner.java:55)
	at org.infinitest.testrunner.TestRunnerProcess.runTest(TestRunnerProcess.java:78)
	at org.infinitest.testrunner.TestRunnerProcess.writeTestResultToOutputStream(TestRunnerProcess.java:137)
	at org.infinitest.testrunner.TestRunnerProcess.main(TestRunnerProcess.java:100)
Caused by: java.lang.IllegalStateException: JMockit didn't get initialized; please check the -javaagent JVM initialization parameter was used
	at mockit.internal.startup.Startup.verifyInitialization(Startup.java:70)
	at mockit.MockUp.<clinit>(MockUp.java:74)
	... 8 more
```
**To Reproduce**

I haven't tried this specifically, but it is hopefully the correct simplification of my code.  Create the following test class:
```
import java.nio.file.Files;
import java.nio.file.Path;

import mockit.Mock;
import mockit.MockUp;

import org.testng.Assert;

public class MissingJmockitParameter
{
   static
   {
      new MockUp<Files>()
      {
         @Mock
         public boolean exists(Path path, LinkOption... options)
         {
            return false;         
         }
      };
   }
   
   @Test
   public static void test()
   {
      Assert.assertTrue(true);
   }
}
```
Run Infinitest and it should reproduce the exception.

**Expected behavior**
Infinitest displays the full exception call stack so the user can root cause the problem.

--- a/infinitest-eclipse/src/main/java/org/infinitest/eclipse/markers/ProblemMarkerInfo.java
+++ b/infinitest-eclipse/src/main/java/org/infinitest/eclipse/markers/ProblemMarkerInfo.java
@@ -39,6 +39,7 @@ import org.eclipse.core.resources.*;
 import org.eclipse.core.runtime.*;
 import org.infinitest.eclipse.workspace.*;
 import org.infinitest.testrunner.*;
+import org.infinitest.testrunner.TestEvent.TestState;
 
 public class ProblemMarkerInfo extends AbstractMarkerInfo {
 	public static final String PICKLED_STACK_TRACE_ATTRIBUTE = "Pickled Stack Trace";
@@ -108,7 +109,13 @@ public class ProblemMarkerInfo extends AbstractMarkerInfo {
 
 	private String buildMessage(TestEvent anEvent) {
 		PointOfFailure failure = anEvent.getPointOfFailure();
-		return anEvent.getErrorClassName() + getMessage(failure) + " in " + stripPackageName(event.getTestName()) + "." + anEvent.getTestMethod();
+		String message = anEvent.getErrorClassName() + getMessage(failure) + " in " + stripPackageName(event.getTestName()) + "." + anEvent.getTestMethod();
+		
+		if (anEvent.getType() == TestState.TEST_INITIALIZATION_FAILURE) {
+			return message + "\n\n" + anEvent.getPrintedStackTrace();
+		} else {
+			return message;
+		}
 	}
 
 	private String getMessage(PointOfFailure failure) {

--- a/infinitest-runner/src/main/java/org/infinitest/TestClassInitializationException.java
+++ b/infinitest-runner/src/main/java/org/infinitest/TestClassInitializationException.java
@@ -1,0 +1,39 @@
+/*
+ * Infinitest, a Continuous Test Runner.
+ *
+ * Copyright (C) 2010-2013
+ * "Ben Rady" <benrady@gmail.com>,
+ * "Rod Coffin" <rfciii@gmail.com>,
+ * "Ryan Breidenbach" <ryan.breidenbach@gmail.com>
+ * "David Gageot" <david@gageot.net>, et al.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package org.infinitest;
+
+/**
+ * 
+ */
+public class TestClassInitializationException extends RuntimeException {
+	private static final long serialVersionUID = 8140038711792655496L;
+
+	public TestClassInitializationException(String message, Throwable t) {
+		super(message, t);
+	}
+}

--- a/infinitest-runner/src/main/java/org/infinitest/testrunner/DefaultRunner.java
+++ b/infinitest-runner/src/main/java/org/infinitest/testrunner/DefaultRunner.java
@@ -28,6 +28,7 @@
 package org.infinitest.testrunner;
 
 import org.infinitest.MissingClassException;
+import org.infinitest.TestClassInitializationException;
 import org.infinitest.config.FileBasedInfinitestConfigurationSource;
 import org.infinitest.config.InfinitestConfigurationSource;
 import org.infinitest.testrunner.junit4.Junit4And3Runner;
@@ -55,6 +56,8 @@ public class DefaultRunner implements NativeRunner {
 			testClass = Class.forName(testClassName);
 		} catch (ClassNotFoundException e) {
 			throw new MissingClassException(testClassName);
+		} catch (Throwable t) {
+			throw new TestClassInitializationException("Failed to initialize test class " + testClassName, t);
 		}
 
 		if (TestNgRunner.isTestNGTest(testClass)) {

--- a/infinitest-runner/src/main/java/org/infinitest/testrunner/TestEvent.java
+++ b/infinitest-runner/src/main/java/org/infinitest/testrunner/TestEvent.java
@@ -32,6 +32,7 @@ import junit.framework.AssertionFailedError;
 import static org.infinitest.testrunner.TestEvent.TestState.*;
 
 import java.io.*;
+import java.nio.charset.StandardCharsets;
 import java.util.Objects;
 
 /**
@@ -43,7 +44,9 @@ public class TestEvent implements Serializable {
 	private static final long serialVersionUID = -1821340797590868333L;
 
 	public enum TestState {
-		METHOD_FAILURE, TEST_CASE_STARTING
+		TEST_INITIALIZATION_FAILURE,
+		METHOD_FAILURE,
+		TEST_CASE_STARTING
 	}
 
 	private final String message;
@@ -52,6 +55,7 @@ public class TestEvent implements Serializable {
 	private final TestState state;
 	private boolean isAssertionFailure;
 	private StackTraceElement[] stackTrace;
+	private String printedStackTrace;
 	private String simpleErrorClassName;
 	private String fullErrorClassName;
 
@@ -73,6 +77,10 @@ public class TestEvent implements Serializable {
 	public static TestEvent methodFailed(String testName, String methodName, Throwable throwable) {
 		return new TestEvent(METHOD_FAILURE, throwable.getMessage(), testName, methodName, throwable);
 	}
+	
+	public static TestEvent testInitializationFailed(String testName, String methodName, Throwable throwable) {
+		return new TestEvent(TEST_INITIALIZATION_FAILURE, throwable.getMessage(), testName, methodName, throwable);
+	}
 
 	public static TestEvent testCaseStarting(String testClass) {
 		return new TestEvent(TEST_CASE_STARTING, "Test Starting", testClass, "", null);
@@ -84,6 +92,17 @@ public class TestEvent implements Serializable {
 		if (stackTrace == null) {
 			stackTrace = new StackTraceElement[0];
 		}
+		
+		try (ByteArrayOutputStream out = new ByteArrayOutputStream();
+				PrintStream stream = new PrintStream(out, false, StandardCharsets.UTF_8)) {
+			error.printStackTrace(stream);
+			stream.flush();
+			
+			printedStackTrace = out.toString(StandardCharsets.UTF_8);
+		} catch (IOException e) {
+			e.printStackTrace();
+		}
+		
 		simpleErrorClassName = error.getClass().getSimpleName();
 		fullErrorClassName = error.getClass().getName();
 	}
@@ -178,5 +197,9 @@ public class TestEvent implements Serializable {
 
 	public StackTraceElement[] getStackTrace() {
 		return stackTrace;
+	}
+	
+	public String getPrintedStackTrace() {
+		return printedStackTrace;
 	}
 }

--- a/infinitest-runner/src/main/java/org/infinitest/testrunner/TestEvent.java
+++ b/infinitest-runner/src/main/java/org/infinitest/testrunner/TestEvent.java
@@ -35,6 +35,8 @@ import java.io.*;
 import java.nio.charset.StandardCharsets;
 import java.util.Objects;
 
+import org.infinitest.MissingClassException;
+
 /**
  * @author <a href="mailto:benrady@gmail.com">Ben Rady</a>
  */
@@ -100,7 +102,7 @@ public class TestEvent implements Serializable {
 			
 			printedStackTrace = out.toString(StandardCharsets.UTF_8);
 		} catch (IOException e) {
-			e.printStackTrace();
+			throw new MissingClassException("Error when trying to print stack trace", e);
 		}
 		
 		simpleErrorClassName = error.getClass().getSimpleName();

--- a/infinitest-runner/src/main/java/org/infinitest/testrunner/TestRunnerProcess.java
+++ b/infinitest-runner/src/main/java/org/infinitest/testrunner/TestRunnerProcess.java
@@ -28,6 +28,7 @@
 package org.infinitest.testrunner;
 
 import static org.infinitest.testrunner.TestEvent.methodFailed;
+import static org.infinitest.testrunner.TestEvent.testInitializationFailed;
 
 import java.io.BufferedReader;
 import java.io.File;
@@ -37,6 +38,7 @@ import java.io.ObjectOutputStream;
 import java.net.Socket;
 import java.nio.charset.StandardCharsets;
 
+import org.infinitest.TestClassInitializationException;
 import org.infinitest.config.FileBasedInfinitestConfigurationSource;
 
 // RISK This class is only tested by running it, which is slow and throws off coverage
@@ -135,13 +137,12 @@ public class TestRunnerProcess {
 		TestResults results;
 		try {
 			results = process.runTest(testName);
-		}
-		// CHECKSTYLE:OFF
-		catch (Throwable e)
-		// CHECKSTYLE:ON
-		{
+		} catch (TestClassInitializationException e) {
+			results = new TestResults(testInitializationFailed(testName, "", e));
+		} catch (Throwable e) {
 			results = new TestResults(methodFailed(testName, "", e));
 		}
+		
 		outputStream.writeObject(results);
 	}
 }

--- a/infinitest-runner/src/main/java/org/infinitest/testrunner/TestRunnerProcess.java
+++ b/infinitest-runner/src/main/java/org/infinitest/testrunner/TestRunnerProcess.java
@@ -133,7 +133,7 @@ public class TestRunnerProcess {
 		}
 	}
 
-	private static void writeTestResultToOutputStream(TestRunnerProcess process, ObjectOutputStream outputStream, String testName) throws IOException {
+	public static void writeTestResultToOutputStream(TestRunnerProcess process, ObjectOutputStream outputStream, String testName) throws IOException {
 		TestResults results;
 		try {
 			results = process.runTest(testName);

--- a/infinitest-runner/src/test/java/org/infinitest/testrunner/TestRunnerProcessTest.java
+++ b/infinitest-runner/src/test/java/org/infinitest/testrunner/TestRunnerProcessTest.java
@@ -27,6 +27,7 @@
  */
 package org.infinitest.testrunner;
 
+import static org.assertj.core.api.Assertions.assertThat;
 import static org.infinitest.config.FileBasedInfinitestConfigurationSource.INFINITEST_FILTERS_FILE_NAME;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertNull;
@@ -37,9 +38,16 @@ import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.verifyNoInteractions;
 
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
 import java.io.File;
 import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
 
+import javax.management.RuntimeErrorException;
+
+import org.infinitest.testrunner.TestEvent.TestState;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.io.TempDir;
 
@@ -99,5 +107,33 @@ class TestRunnerProcessTest {
 	void getFilterFile() {
 		File filteFile = new File(INFINITEST_FILTERS_FILE_NAME);
 		assertEquals(filteFile, TestRunnerProcess.getFilterFile(new String[] {null, null, INFINITEST_FILTERS_FILE_NAME}));
+	}
+	
+	@Test
+	void testClassInitializationFailure() throws IOException, ClassNotFoundException {
+		try (ByteArrayOutputStream out = new ByteArrayOutputStream(); ObjectOutputStream oos = new ObjectOutputStream(out)) {
+			TestRunnerProcess process = new TestRunnerProcess(DefaultRunner.class.getName(), null);
+			TestRunnerProcess.writeTestResultToOutputStream(process, oos, InvalidTestClass.class.getName());
+			
+			try (ByteArrayInputStream in = new ByteArrayInputStream(out.toByteArray()); ObjectInputStream ois = new ObjectInputStream(in)) {
+				TestResults results = (TestResults) ois.readObject();
+				
+				assertThat(results.iterator().hasNext()).isTrue();
+				
+				TestEvent event = results.iterator().next();
+				
+				assertThat(event.getType()).isEqualTo(TestState.TEST_INITIALIZATION_FAILURE);
+				assertThat(event.getPrintedStackTrace()).isNotNull();
+			}
+		}
+	}
+	
+	private static class InvalidTestClass {
+		static {
+			Exception cause = new Exception("Test for a class failing to load");
+			if (true) {
+				throw new RuntimeException("Failed to load class", cause);
+			}
+		}
 	}
 }

--- a/infinitest-runner/src/test/java/org/infinitest/testrunner/TestRunnerProcessTest.java
+++ b/infinitest-runner/src/test/java/org/infinitest/testrunner/TestRunnerProcessTest.java
@@ -45,8 +45,6 @@ import java.io.IOException;
 import java.io.ObjectInputStream;
 import java.io.ObjectOutputStream;
 
-import javax.management.RuntimeErrorException;
-
 import org.infinitest.testrunner.TestEvent.TestState;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.io.TempDir;

